From 8eef36a105a34244e5594ad5fb0f688c0e40ced4 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Tue, 22 Sep 2020 12:24:21 -0700
Subject: [PATCH 1646/1852] v4l: xilinx: cresample: Fix warnings for
 NULL_RETURNS

commit 86167ca93c37e93d1fb0ea1c77a4b4763d313b7f from
https://github.com/Xilinx/linux-xlnx.git

This patch fix the warnings for the NULL_RETURNS type.

The actual warnings are:

For __xcresample_get_pad_format():
Event return_null: 	Explicitly returning null.

For xcresample_get_format():
Event returned_null: 	"__xcresample_get_pad_format" returns "NULL".
Event dereference: 	Dereferencing
"__xcresample_get_pad_format(xcresample, cfg, fmt->pad, fmt->which)",
which is known to be "NULL".

For xcresample_set_format():
Event returned_null: 	"__xcresample_get_pad_format" returns "NULL".
Event var_assigned: 	Assigning: "format" = "NULL" return value from
"__xcresample_get_pad_format".
Event dereference: 	Dereferencing "format", which is known to be
"NULL"

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../media/platform/xilinx/xilinx-cresample.c  | 26 ++++++++++++++-----
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-cresample.c b/drivers/media/platform/xilinx/xilinx-cresample.c
index 05335c10a388..ae04c0fb1c47 100644
--- a/drivers/media/platform/xilinx/xilinx-cresample.c
+++ b/drivers/media/platform/xilinx/xilinx-cresample.c
@@ -89,15 +89,22 @@ __xcresample_get_pad_format(struct xcresample_device *xcresample,
 			    struct v4l2_subdev_pad_config *cfg,
 			    unsigned int pad, u32 which)
 {
+	struct v4l2_mbus_framefmt *format;
+
 	switch (which) {
 	case V4L2_SUBDEV_FORMAT_TRY:
-		return v4l2_subdev_get_try_format(&xcresample->xvip.subdev, cfg,
-						  pad);
+		format = v4l2_subdev_get_try_format(&xcresample->xvip.subdev,
+						    cfg, pad);
+		break;
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xcresample->formats[pad];
+		format = &xcresample->formats[pad];
+		break;
 	default:
-		return NULL;
+		format = NULL;
+		break;
 	}
+
+	return format;
 }
 
 static int xcresample_get_format(struct v4l2_subdev *subdev,
@@ -105,9 +112,14 @@ static int xcresample_get_format(struct v4l2_subdev *subdev,
 				 struct v4l2_subdev_format *fmt)
 {
 	struct xcresample_device *xcresample = to_cresample(subdev);
+	struct v4l2_mbus_framefmt *format;
 
-	fmt->format = *__xcresample_get_pad_format(xcresample, cfg, fmt->pad,
-						   fmt->which);
+	format = __xcresample_get_pad_format(xcresample, cfg, fmt->pad,
+					     fmt->which);
+	if (!format)
+		return -EINVAL;
+
+	fmt->format = *format;
 
 	return 0;
 }
@@ -121,6 +133,8 @@ static int xcresample_set_format(struct v4l2_subdev *subdev,
 
 	format = __xcresample_get_pad_format(xcresample, cfg, fmt->pad,
 					     fmt->which);
+	if (!format)
+		return -EINVAL;
 
 	if (fmt->pad == XVIP_PAD_SOURCE) {
 		fmt->format = *format;
-- 
2.31.1

